lib/pact_broker/index/service.rb

```diff
@@ -194,6 +184,39 @@ def self.latest_verifications_for_consumer_version_tags(options)
           nil # should not be used
         end
       end
+
+      def self.base_query(options)
+        query = PactBroker::Pacts::PactPublication.select(Sequel[:pact_publications][:id])
+
+        if options[:consumer_name]
+          consumer = pacticipant_repository.find_by_name!(options[:consumer_name])
+          query = query.for_consumer(consumer)
+        end
+
+        if options[:provider_name]
+          provider = pacticipant_repository.find_by_name!(options[:provider_name])
+          query = query.for_provider(provider)
+        end
+
+        query
+      end
+
+      def self.pact_publication_ids_query(query, tags)
+        return query.union(query.latest_for_consumer_tag(tags)) if tags.is_a?(Array)
+        return query.union(query.latest_by_consumer_tag) if tags
+
+        query
+      end
+
+      def self.pacticipant_name_query(pacticipant_name)
```

Comment:
This is getting closer! But it's still filtering a really big slow dataset (the entire index query) on a string match, rather than finding the pacticipants (a really small query) on a string match, and then being able to set the exact IDs of the consumer and provider in the big query (which will make it super quick). What's your thinking behind that?

Reply:
> What's your thinking behind that?

I was thinking doing 2 round trips to the database is a bit inefficient.
However, I think I'm wrong as you said filtering string match with `ilike` is also an expensive query

I'll do it the way you suggested and probably find the way to optimise the current index query on a different issue/PR :)

Thanks for being patient with me :)

Reply:
Attemp in 8fd82a57

The new SQL query looks like below
```sql
SELECT pact_publications.*
FROM pact_publications
INNER JOIN pacticipants AS consumers ON (pact_publications.consumer_id = consumers.id)
INNER JOIN pacticipants AS providers ON (pact_publications.provider_id = providers.id)
INNER JOIN versions AS cv ON (pact_publications.consumer_version_id = cv.id)
WHERE ((pact_publications.id IN
          (SELECT p.*
           FROM
             (SELECT pact_publications.id
              FROM pact_publications
              INNER JOIN versions AS cv ON (pact_publications.consumer_version_id = cv.id)
              LEFT JOIN
                (SELECT consumer_id,
                        provider_id,
                        "order"
                 FROM pact_publications
                 INNER JOIN versions AS cv ON (pact_publications.consumer_version_id = cv.id)) AS pp2 ON ((pact_publications.consumer_id = pp2.consumer_id)
                                                                                                          AND (pact_publications.provider_id = pp2.provider_id)
                                                                                                          AND (pp2.order > cv.order))
              WHERE (pp2.order IS NULL)) AS p
           INNER JOIN latest_pact_publication_ids_for_consumer_versions AS lp ON (lp.pact_publication_id = p.id)))
       AND (pact_publications.id IN
              (SELECT pact_publications.id
               FROM pact_publications
               INNER JOIN pacticipants AS consumers ON (pact_publications.consumer_id = consumers.id)
               INNER JOIN pacticipants AS providers ON (pact_publications.provider_id = providers.id)
               WHERE ((providers.name ILIKE '%test%' ESCAPE '\')
                      OR (providers.name ILIKE '%ap\_tesp%' ESCAPE '\')
                      OR (consumers.name ILIKE '%test\_d%' ESCAPE '\')
                      OR (consumers.name ILIKE '%appdf%' ESCAPE '\')))))
```

Reply:
Please let me know how you think.
There is a failed test in `app_spec.rb` which Im a bit baffled to understand. Will look at it later after you are happy with the filtering :)

Reply:
This is the part that gets added to a big index query

```sql
AND (pact_publications.id IN
              (SELECT pact_publications.id
               FROM pact_publications
               INNER JOIN pacticipants AS consumers ON (pact_publications.consumer_id = consumers.id)
               INNER JOIN pacticipants AS providers ON (pact_publications.provider_id = providers.id)
               WHERE ((providers.name ILIKE '%test%' ESCAPE '\')
                      OR (providers.name ILIKE '%ap\_tesp%' ESCAPE '\')
                      OR (consumers.name ILIKE '%test\_d%' ESCAPE '\')
                      OR (consumers.name ILIKE '%appdf%' ESCAPE '\')))))
```

Reply:
The problem with this approach is that it has to do all the joins, creating a big dataset, then remove lines by filtering on the pacticipant names afterwards. 

By doing the query to get the pacticipant ids first, and only joining those consumers, then the dataset never gets big in the first place. Ideally, we want a query that looks like:
```
SELECT pact_publications.id
               FROM pact_publications
               INNER JOIN pacticipants AS consumers ON (pact_publications.consumer_id = consumers.id and pact_publications.consumer_id in [<the ids from the first query>])
               INNER JOIN pacticipants AS providers ON (pact_publications.provider_id = providers.id and providers.id in [<the ids from the first query>])
```

I've learned that making an extra round trip to the database is worth the trade off when you can significantly cut down on the amount of data being processed in a subsequent query.

Reply:
Sorry, I put those IN statements in the wrong part of the query.


```
SELECT pact_publications.*
FROM pact_publications
INNER JOIN pacticipants AS consumers ON (pact_publications.consumer_id = consumers.id)
INNER JOIN pacticipants AS providers ON (pact_publications.provider_id = providers.id)
INNER JOIN versions AS cv ON (pact_publications.consumer_version_id = cv.id)
WHERE ((pact_publications.id IN
          (SELECT p.*
           FROM
             (SELECT pact_publications.id
              FROM pact_publications
WHERE ( pact_publications.consumer_id IN [...]  OR pact_publications.provider_id IN [...])
              INNER JOIN versions AS cv ON (pact_publications.consumer_version_id = cv.id)
              LEFT JOIN
                (SELECT consumer_id,
                        provider_id,
                        "order"
                 FROM pact_publications
                 INNER JOIN versions AS cv ON (pact_publications.consumer_version_id = cv.id)) AS pp2 ON ((pact_publications.consumer_id = pp2.consumer_id)
                                                                                                          AND (pact_publications.provider_id = pp2.provider_id)
                                                                                                          AND (pp2.order > cv.order))
              WHERE (pp2.order IS NULL)) AS p
           INNER JOIN latest_pact_publication_ids_for_consumer_versions AS lp ON (lp.pact_publication_id = p.id)))
       AND (pact_publications.id IN
              (SELECT pact_publications.id
               FROM pact_publications
WHERE ( pact_publications.consumer_id IN [...]  OR pact_publications.provider_id IN [...])
               INNER JOIN pacticipants AS consumers ON (pact_publications.consumer_id = consumers.id)
               INNER JOIN pacticipants AS providers ON (pact_publications.provider_id = providers.id)
)
```

See how that's filtering the base table before it gets joined, rather than on a joined table after it gets joined?